src/ray/gcs/gcs_server/gcs_placement_group_scheduler.cc

```diff
@@ -0,0 +1,241 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "ray/gcs/gcs_server/gcs_placement_group_scheduler.h"
+#include <ray/protobuf/gcs.pb.h>
+#include "ray/gcs/gcs_server/gcs_placement_group_manager.h"
+
+namespace ray {
+namespace gcs {
+
+GcsPlacementGroupScheduler::GcsPlacementGroupScheduler(
+    boost::asio::io_context &io_context,
+    std::shared_ptr<gcs::GcsTableStorage> gcs_table_storage,
+    const gcs::GcsNodeManager &gcs_node_manager,
+    std::function<void(std::shared_ptr<GcsPlacementGroup>)> schedule_failure_handler,
+    std::function<void(std::shared_ptr<GcsPlacementGroup>)> schedule_success_handler,
+    ReserveResourceClientFactoryFn lease_client_factory)
+    : io_context_(io_context),
+      gcs_table_storage_(gcs_table_storage),
+      gcs_node_manager_(gcs_node_manager),
+      schedule_failure_handler_(std::move(schedule_failure_handler)),
+      schedule_success_handler_(std::move(schedule_success_handler)),
+      lease_client_factory_(std::move(lease_client_factory)) {
+  RAY_CHECK(schedule_failure_handler_ != nullptr && schedule_success_handler_ != nullptr);
+  scheduler.push_back(std::make_shared<GcsPackStrategy>());
+  scheduler.push_back(std::make_shared<GcsSpreadStrategy>());
+}
+
+std::unordered_map<BundleID, ClientID> GcsPackStrategy::Schedule(
+    std::vector<ray::BundleSpecification> &bundles, const GcsNodeManager &node_manager) {
+  std::unordered_map<BundleID, ClientID> schedule_map;
+  auto &alive_nodes = node_manager.GetAllAliveNodes();
+  for (size_t pos = 0; pos < bundles.size(); pos++) {
+    schedule_map[bundles[pos].BundleId()] =
+        ClientID::FromBinary(alive_nodes.begin()->second->node_id());
+  }
+  return schedule_map;
+}
+
+std::unordered_map<BundleID, ClientID> GcsSpreadStrategy::Schedule(
+    std::vector<ray::BundleSpecification> &bundles, const GcsNodeManager &node_manager) {
+  std::unordered_map<BundleID, ClientID> schedule_map;
+  auto &alive_nodes = node_manager.GetAllAliveNodes();
+  auto iter = alive_nodes.begin();
+  size_t index = 0;
+  size_t alive_nodes_size = alive_nodes.size();
+  for (; iter != alive_nodes.end(); iter++, index++) {
+    for (size_t base = 0;; base++) {
+      if (index + base * alive_nodes_size >= bundles.size()) {
+        break;
+      } else {
+        schedule_map[bundles[index + base * alive_nodes_size].BundleId()] =
+            ClientID::FromBinary(iter->second->node_id());
+      }
+    }
+  }
+  return schedule_map;
+}
+
+void GcsPlacementGroupScheduler::Schedule(
```

Comment:
Can you add some overview comment of the scheduling strategy? It seems like you start with some initial decision and then try to refine it?

Reply:
I add some  comments now.